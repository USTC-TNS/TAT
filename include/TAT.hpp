/* TAT.hpp
 * Copyright (C) 2019  Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef TAT_HPP_
#define TAT_HPP_

#ifndef TAT_VERSION
#define TAT_VERSION "unknown"
#endif // TAT_VERSION

#include <cassert>
#include <cstring>
#include <cmath>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <memory>
#include <functional>
#include <complex>

#define PASS std::cerr << "calling a passing function at " << __FILE__ << ":" << __LINE__ << " in " << __PRETTY_FUNCTION__ << std::endl, exit(233)
#define ENABLE_IF(...) class = typename std::enable_if<__VA_ARGS__::value>::type

#if (!defined TAT_USE_CPU && !defined TAT_USE_CUDA && !defined TAT_USE_DCU && !defined TAT_USE_SW)
#if !defined TAT_DEFAULT
#warning use CPU by default
#endif
#define TAT_USE_CPU
#endif

#ifdef TAT_EXTREME
#warning EXTREME compile may cost much of compile time
#endif // TAT_EXTREME

#ifdef TAT_USE_CPU
extern "C"
{
#define MKL_Complex8 std::complex<float>
#define MKL_Complex16 std::complex<double>
#include <mkl.h>
} // extern "C"
#include <hptt.h>
#ifdef TAT_EXTREME
#include <../src/hptt.cpp>
#include <../src/plan.cpp>
#include <../src/transpose.cpp>
#include <../src/utils.cpp>
#endif // TAT_EXTREME
#include <rang.hpp>

// SVD
#if (defined TAT_USE_GESDD && defined TAT_USE_GESVD) || (defined TAT_USE_GESVD && defined TAT_USE_GESVDX) || (defined TAT_USE_GESVDX && defined TAT_USE_GESDD)
#error only one of GESDD, GESVD and GESVDX could be in use
#endif
#if (!defined TAT_USE_GESDD && !defined TAT_USE_GESVD && !defined TAT_USE_GESVDX)
#if !defined TAT_DEFAULT
#warning must use one of GESDD, GESVD and GESVDX, default use GESVD now
#endif
#define TAT_USE_GESVD
#endif

// QR
#if (defined TAT_USE_GEQRF && defined TAT_USE_GEQP3)
#error only one of GEQRF and GEQP3 could be in use
#endif
#if (!defined TAT_USE_GEQRF && !defined TAT_USE_GEQP3)
#if !defined TAT_DEFAULT
#warning must use one of GEQRF and GEQP3, default use GEQRF now
#endif
#define TAT_USE_GEQRF
#endif

#ifdef TAT_USE_GEQP3
#error GEQP3 is current unusable
#endif

#endif // TAT_USE_CPU

namespace TAT {
  template<class T>
  class is_scalar {
   public:
    static constexpr bool value = std::is_scalar<T>::value || std::is_same<T, std::complex<float>>::value || std::is_same<T, std::complex<double>>::value || std::is_same<T, std::complex<long double>>::value;
  };

  template<class Base>
  class RealBaseClass {
   public:
    using type=Base;
  };
  template<>
  class RealBaseClass<std::complex<float>> {
   public:
    using type=float;
  };
  template<>
  class RealBaseClass<std::complex<double>> {
   public:
    using type=double;
  };
  template<>
  class RealBaseClass<std::complex<long double>> {
   public:
    using type=long double;
  };
  template<class T>
  using RealBase = typename RealBaseClass<T>::type;

  enum class Device : unsigned char {CPU, CUDA, DCU, SW};

  namespace legs {
    class Legs {
     public:
      unsigned char id;
      bool operator==(const Legs& other) const {
        return id==other.id;
      }
      bool operator!=(const Legs& other) const {
        return id!=other.id;
      }
      bool operator<(const Legs& other) const {
        return id<other.id;
      }
      Legs() = default;
      Legs(const std::string& name) {
        try {
          id = name2id.at(name);
        } catch (const std::out_of_range& e) {
          id = total++;
          name2id[name]=id;
          id2name[id]=name;
        } // exsit name
      }
      static unsigned char total;
      static std::map<std::string, unsigned char> name2id;
      static std::map<unsigned char, std::string> id2name;

      friend std::ostream& operator<<(std::ostream& out, const Legs& value) {
        try {
          return out << id2name.at(value.id);
        } catch (const std::out_of_range& e) {
          return out << "UserDefinedLeg" << value.id;
        } // get name
      } // operator<<
    }; // class Legs

    unsigned char Legs::total = 0;
    std::map<std::string, unsigned char> Legs::name2id = {};
    std::map<unsigned char, std::string> Legs::id2name = {};
  } // namespace legs
  using legs::Legs;

  namespace legs_name {
#define TAT_DefineLeg(x) static const TAT::Legs x(#x)
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Phy##n); TAT_DefineLeg(Left##n); TAT_DefineLeg(Right##n); TAT_DefineLeg(Up##n); TAT_DefineLeg(Down##n); \
      TAT_DefineLeg(LeftUp##n); TAT_DefineLeg(LeftDown##n); TAT_DefineLeg(RightUp##n); TAT_DefineLeg(RightDown##n)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Tmp##n##0); TAT_DefineLeg(Tmp##n##1); TAT_DefineLeg(Tmp##n##2); TAT_DefineLeg(Tmp##n##3); TAT_DefineLeg(Tmp##n##4); \
      TAT_DefineLeg(Tmp##n##5); TAT_DefineLeg(Tmp##n##6); TAT_DefineLeg(Tmp##n##7); TAT_DefineLeg(Tmp##n##8); TAT_DefineLeg(Tmp##n##9)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#undef TAT_DefineLeg
    // total = 190 currently
  } // namespace legs_name

  using Size = std::size_t;
  using Rank = unsigned int;

  namespace data {
    template<Device device, class Base>
    class Magic;
#define DefineData(x) \
      namespace x { \
        template<class Base> \
        class Data; \
      } \
      template<class Base> \
      class Magic<Device::x, Base>{ \
       public: \
        using type=x::Data<Base>; \
      }

    DefineData(CPU);
    DefineData(CUDA);
    DefineData(DCU);
    DefineData(SW);
#undef DefineData

    template<Device device=Device::CPU, class Base=double, ENABLE_IF(is_scalar<Base>)>
    using Data = typename Magic<device, Base>::type;
  } // namespace data
  using data::Data;

  namespace block {
    template<Device device=Device::CPU, class Base=double>
    class Block;
  } // namespace block
  using block::Block;

  namespace tensor {
    template<Device device=Device::CPU, class Base=double>
    class Tensor;
  } // namespace tensor
  //using tensor::Tensor;
  template<Device device=Device::CPU, class Base=double>
  using Tensor=Block<device, Base>;

  namespace node {
    template<Device device=Device::CPU, class Base=double>
    class Node;
  } // namespace node
  using node::Node;

  namespace lazy {
    template<Device device=Device::CPU, class Base=double>
    class Lazy;
  } // namespace lazy
  using lazy::Lazy;

  namespace site {
    template<Device device=Device::CPU, class Base=double>
    class Site;
  } // namespace site
  using site::Site;

  namespace lattice {
    template<int dimension>
    class Dimension;

    template<class Tags=Dimension<2>, Device device=Device::CPU, class Base=double>
    class Lattice;
  } // namespace lattice
  using site::Site;
} // namespace TAT

#include "TAT/Data.hpp"
#include "TAT/Block.hpp"
// Tensor
#include "TAT/Node.hpp"
#include "TAT/Lazy.hpp"

#include "TAT/Site.hpp"
#include "TAT/Lattice.hpp"

#endif // TAT_HPP_
