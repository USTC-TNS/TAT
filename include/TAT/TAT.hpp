/**
 * \file TAT.hpp
 *
 * Copyright (C) 2019-2020 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_HPP
#define TAT_HPP

#ifndef __cplusplus
#error only work for c++
#endif

#ifdef _MSVC_LANG
#if _MSVC_LANG < 201703L
#error require c++17 or later
#endif
#else
#if __cplusplus < 201703L
#error require c++17 or later
#endif
#endif

#ifdef _WIN32
#include <windows.h>
#endif

// 开关说明
// TAT_USE_MPI 定义以开启MPI支持, cmake可对此进行定义
// TAT_USE_SIMPLE_NAME 定义以使用原始字符串作为name

namespace TAT {
   /**
    * \brief TAT的版本号
    */
   inline const char* version = "0.0.5";

   /**
    * \brief Debug模式中, 将在程序末尾打印一行友情提示, 过早的优化是万恶之源, 同时控制windows下终端的色彩模式
    */
   struct evil_t {
      evil_t() {
#ifdef _WIN32
         HANDLE output_handle = GetStdHandle(STD_OUTPUT_HANDLE);
         DWORD output_mode = 0;
         GetConsoleMode(output_handle, &output_mode);
         SetConsoleMode(output_handle, output_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
         HANDLE error_handle = GetStdHandle(STD_ERROR_HANDLE);
         DWORD error_mode = 0;
         GetConsoleMode(error_handle, &error_mode);
         SetConsoleMode(error_handle, error_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
#endif
      }
      ~evil_t();
   };
   inline const evil_t evil;
} // namespace TAT

// clang-format off
#include "tensor.hpp"
#include "implement.hpp"
#include "tools.hpp"
// clang-format on

#endif
