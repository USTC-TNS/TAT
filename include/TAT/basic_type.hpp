/**
 * \file basic_type.hpp
 *
 * Copyright (C) 2019-2020 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_BASIC_TYPE_HPP
#define TAT_BASIC_TYPE_HPP

#include <complex>
#include <type_traits>

namespace TAT {
   /**
    * \brief 张量的秩的类型
    */
   using Rank = unsigned short;
   /**
    * \brief 张量数据维度大小和数据本身大小的类型
    */
   using Size = unsigned long;

   /**
    * \brief Z2对称性的类型
    */
   using Z2 = bool;
   /**
    * \brief U1对称性的类型
    */
   using U1 = int;
   /**
    * \brief 费米子数目的类型
    */
   using Fermi = short;

   /**
    * \brief 费米箭头方向的类型, false和true分别表示出入
    */
   using Arrow = bool;

   /**
    * \brief 判断一个类型是否是标量类型, 修复了std::scalar不能判断std::complex的问题
    * \tparam T 如果T是标量类型, 则value为true
    * \see is_scalar_v
    */
   template<typename T>
   struct is_scalar : std::is_scalar<T> {};
   template<typename T>
   struct is_scalar<std::complex<T>> : std::is_scalar<T> {};
   template<typename T>
   constexpr bool is_scalar_v = is_scalar<T>::value;

   /**
    * \brief 取对应的实数类型, 在svd, norm等地方会用到
    * \tparam T 如果T是std::complex<S>, 则type为S, 若T为其他标量类型, 则type为T本身, 否则为void
    * \see real_base_t
    */
   template<typename T>
   struct real_base : std::conditional<is_scalar<T>::value, T, void> {};
   template<typename T>
   struct real_base<std::complex<T>> : std::conditional<is_scalar<T>::value, T, void> {};
   template<typename T>
   using real_base_t = typename real_base<T>::type;

   /**
    * \brief 判断是否是复数类型
    * \see is_complex_v
    */
   template<typename T>
   struct is_complex : std::is_same<T, std::complex<real_base_t<T>>> {};
   template<typename T>
   constexpr bool is_complex_v = is_complex<T>::value;
   /**
    * \brief 判断是否是实数类型
    * \see is_real_v
    */
   template<typename T>
   struct is_real : std::is_same<T, real_base_t<T>> {};
   template<typename T>
   constexpr bool is_real_v = is_real<T>::value;
} // namespace TAT
#endif
