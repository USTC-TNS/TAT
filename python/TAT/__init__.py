import cppyy
cppyy.load_library("blas")
cppyy.load_library("lapack")
cppyy.include("../include/TAT/TAT.hpp")
std = cppyy.gbl.std
TAT = cppyy.gbl.TAT
