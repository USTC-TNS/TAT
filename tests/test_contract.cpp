#include <TAT/TAT.hpp>
#include <gmock/gmock.h>
#include <gtest/gtest.h>

using namespace testing;

TEST(test_contract, no_symmetry_example_0) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B"}, {2, 2}}.range();
    auto b = TAT::Tensor<double, TAT::NoSymmetry>{{"C", "D"}, {2, 2}}.range();

    auto c = TAT::Tensor<double, TAT::NoSymmetry>::contract(a, b, {{"A", "C"}});
    auto c_expect =
        TAT::Tensor<double, TAT::NoSymmetry>{{"B", "D"}, {2, 2}}.set([i = 0, v = std::vector<double>{4, 6, 6, 10}]() mutable { return v[i++]; });
    ASSERT_FLOAT_EQ((c - c_expect).norm<-1>(), 0);

    auto d = TAT::Tensor<double, TAT::NoSymmetry>::contract(a, b, {{"A", "D"}});
    auto d_expect =
        TAT::Tensor<double, TAT::NoSymmetry>{{"B", "C"}, {2, 2}}.set([i = 0, v = std::vector<double>{2, 6, 3, 11}]() mutable { return v[i++]; });
    ASSERT_FLOAT_EQ((d - d_expect).norm<-1>(), 0);

    auto e = TAT::Tensor<double, TAT::NoSymmetry>::contract(a, b, {{"B", "C"}});
    auto e_expect =
        TAT::Tensor<double, TAT::NoSymmetry>{{"A", "D"}, {2, 2}}.set([i = 0, v = std::vector<double>{2, 3, 6, 11}]() mutable { return v[i++]; });
    ASSERT_FLOAT_EQ((e - e_expect).norm<-1>(), 0);

    auto f = TAT::Tensor<double, TAT::NoSymmetry>::contract(a, b, {{"B", "D"}});
    auto f_expect =
        TAT::Tensor<double, TAT::NoSymmetry>{{"A", "C"}, {2, 2}}.set([i = 0, v = std::vector<double>{1, 3, 3, 13}]() mutable { return v[i++]; });
    ASSERT_FLOAT_EQ((f - f_expect).norm<-1>(), 0);
}

TEST(test_contract, no_symmetry_example_1) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>::contract(
        TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B", "C", "D"}, {1, 2, 3, 4}}.range(),
        TAT::Tensor<double, TAT::NoSymmetry>{{"E", "F", "G", "H"}, {3, 1, 2, 4}}.range(),
        {{"B", "G"}, {"D", "H"}}
    );
    auto b = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "C", "E", "F"}, {1, 3, 3, 1}}.set(
        [i = 0, v = std::vector<double>{316, 796, 1276, 428, 1164, 1900, 540, 1532, 2524}]() mutable { return v[i++]; }
    );
    ASSERT_FLOAT_EQ((a - b).norm<-1>(), 0);
}

TEST(test_contract, u1_symmetry_example_0) {
    using Tensor = TAT::Tensor<double, TAT::U1Symmetry>;
    auto edge1 = TAT::Edge<TAT::U1Symmetry>{{-1, 2}, {0, 2}, {+1, 2}};
    auto edge2 = TAT::Edge<TAT::U1Symmetry>{{+1, 2}, {0, 2}, {-1, 2}};
    auto a = Tensor({"a", "b", "c", "d"}, {edge1, edge2, edge1, edge2}).range();
    auto b = Tensor({"e", "f", "g", "h"}, {edge1, edge2, edge1, edge2}).range();
    for (auto plan : std::vector<std::unordered_set<std::pair<std::string, std::string>>>{
             {{"a", "f"}, {"b", "e"}}, {{"a", "f"}, {"b", "g"}}, {{"a", "h"}, {"b", "e"}}, {{"a", "h"}, {"b", "g"}}, {{"a", "f"}, {"c", "h"}},
             {{"a", "h"}, {"c", "f"}}, {{"a", "f"}, {"d", "e"}}, {{"a", "f"}, {"d", "g"}}, {{"a", "h"}, {"d", "e"}}, {{"a", "h"}, {"d", "g"}},
             {{"c", "f"}, {"b", "e"}}, {{"c", "f"}, {"b", "g"}}, {{"c", "h"}, {"b", "e"}}, {{"c", "h"}, {"b", "g"}}, {{"b", "e"}, {"d", "g"}},
             {{"b", "g"}, {"d", "e"}}, {{"c", "f"}, {"d", "e"}}, {{"c", "f"}, {"d", "g"}}, {{"c", "h"}, {"d", "e"}}, {{"c", "h"}, {"d", "g"}},
         }) {
        auto c = a.contract(b, plan).clear_symmetry();
        auto d = a.clear_symmetry().contract(b.clear_symmetry(), plan);
        ASSERT_FLOAT_EQ((c - d).norm<-1>(), 0);
    }
}

TEST(test_contract, fermi_symmetry_example_0) {
    using FermiTensor = TAT::Tensor<double, TAT::FermiSymmetry>;
    auto fermi_edge1 = TAT::edge_segments_t<TAT::FermiSymmetry>({{-1, 2}, {0, 2}, {+1, 2}});
    auto fermi_edge2 = TAT::edge_segments_t<TAT::FermiSymmetry>({{+1, 2}, {0, 2}, {-1, 2}});
    auto fermi_a = FermiTensor({"a", "b", "c", "d"}, {{fermi_edge1, true}, fermi_edge2, {fermi_edge1, true}, {fermi_edge2, true}}).range();
    auto fermi_b = FermiTensor({"e", "f", "g", "h"}, {fermi_edge1, fermi_edge2, {fermi_edge1, true}, fermi_edge2}).range();
    auto fermi_c = fermi_a.contract(fermi_b, {{"d", "e"}, {"c", "f"}});
    auto fermi_d = fermi_b.contract(fermi_a, {{"e", "d"}, {"f", "c"}});
    ASSERT_FLOAT_EQ((fermi_c - fermi_d).norm<-1>(), 0);

    using U1Tensor = TAT::Tensor<double, TAT::U1Symmetry>;
    auto u1_edge1 = TAT::edge_segments_t<TAT::U1Symmetry>({{-1, 2}, {0, 2}, {+1, 2}});
    auto u1_edge2 = TAT::edge_segments_t<TAT::U1Symmetry>({{+1, 2}, {0, 2}, {-1, 2}});
    auto u1_a = U1Tensor({"a", "b", "c", "d"}, {u1_edge1, u1_edge2, u1_edge1, u1_edge2}).range();
    auto u1_b = U1Tensor({"e", "f", "g", "h"}, {u1_edge1, u1_edge2, u1_edge1, u1_edge2}).range();
    auto u1_c = u1_a.contract(u1_b, {{"d", "e"}, {"c", "f"}});

    ASSERT_THAT(fermi_a.storage(), ElementsAreArray(u1_a.storage()));
    ASSERT_THAT(fermi_b.storage(), ElementsAreArray(u1_b.storage()));
    ASSERT_THAT(fermi_c.storage(), ElementsAreArray(u1_c.storage()));
}

TEST(test_contract, contract_with_split_and_merge) {
    using Tensor = TAT::Tensor<double, TAT::FermiSymmetry>;
    auto edge1 = TAT::Edge<TAT::FermiSymmetry>({{-1, 2}, {0, 2}, {+1, 2}}, false);
    auto edge2 = TAT::Edge<TAT::FermiSymmetry>({{+1, 2}, {0, 2}, {-1, 2}}, true);
    auto a = Tensor({"a", "b", "c", "d"}, {edge1, edge2, edge1, edge2}).range();
    auto b = Tensor({"e", "f", "g", "h"}, {edge1, edge2, edge1, edge2}).range();
    auto c = a.contract(b, {{"a", "f"}, {"b", "g"}, {"c", "h"}});

    auto a_merged = a.merge_edge({{"m", {"b", "a"}}}, false);
    auto b_merged = b.merge_edge({{"m", {"g", "f"}}}, true);
    auto c_merged = a_merged.contract(b_merged, {{"m", "m"}, {"c", "h"}});

    ASSERT_FLOAT_EQ((c - c_merged).norm<-1>(), 0);
}

TEST(test_contract, contract_with_reverse_0) {
    auto a = TAT::Tensor<double, TAT::ParitySymmetry>({"i", "j"}, {{{{false, 2}, {true, 2}}, false}, {{{false, 2}, {true, 2}}, true}}).range();
    auto b = TAT::Tensor<double, TAT::ParitySymmetry>({"i", "j"}, {{{{false, 2}, {true, 2}}, false}, {{{false, 2}, {true, 2}}, true}})
                 .range()
                 .transpose({"j", "i"});
    auto c = a.contract(b, {{"j", "i"}});

    auto a_reversed = a.reverse_edge({"j"}, false);
    auto b_reversed = b.reverse_edge({"i"}, true);
    auto c_reversed = a_reversed.contract(b_reversed, {{"j", "i"}});

    ASSERT_FLOAT_EQ((c - c_reversed).norm<-1>(), 0);
}

TEST(test_contract, contract_with_reverse_1) {
    using Tensor = TAT::Tensor<double, TAT::FermiSymmetry>;
    auto edge1 = TAT::Edge<TAT::FermiSymmetry>({{-1, 2}, {0, 2}, {+1, 2}}, false);
    auto edge2 = TAT::Edge<TAT::FermiSymmetry>({{+1, 2}, {0, 2}, {-1, 2}}, true);
    auto a = Tensor({"a", "b", "c", "d"}, {edge1, edge2, edge1, edge2}).range();
    auto b = Tensor({"e", "f", "g", "h"}, {edge1, edge2, edge1, edge2}).range();
    auto c = a.contract(b, {{"a", "f"}, {"b", "g"}, {"c", "h"}});

    auto a_reversed = a.reverse_edge({"b", "a"}, false);
    auto b_reversed = b.reverse_edge({"g", "f"}, true);
    auto c_reversed = a_reversed.contract(b_reversed, {{"a", "f"}, {"b", "g"}, {"c", "h"}});

    ASSERT_FLOAT_EQ((c - c_reversed).norm<-1>(), 0);
}

TEST(test_contract, fuse) {
    TAT::Size fuse_d = 3;
    TAT::Size common_d = 4;
    auto a = TAT::Tensor<>({"A", "B", "C"}, {fuse_d, common_d, 5}).range();
    auto b = TAT::Tensor<>({"A", "B", "D"}, {fuse_d, common_d, 7}).range();
    auto c = TAT::Tensor<>::contract(a, b, {{"B", "B"}}, {"A"});
    for (TAT::Size i = 0; i < fuse_d; i++) {
        auto hat = TAT::Tensor<>({"A"}, {fuse_d}).zero();
        hat.storage()[i] = 1;
        auto a0 = a.contract(hat, {{"A", "A"}});
        auto b0 = b.contract(hat, {{"A", "A"}});
        auto c0 = c.contract(hat, {{"A", "A"}});
        ASSERT_FLOAT_EQ((a0.contract(b0, {{"B", "B"}}) - c0).norm<-1>(), 0);
    }
}

TEST(test_contract, corner_no_symmetry_0k) {
    auto a = TAT::Tensor<double, TAT::NoSymmetry>{{"A", "B"}, {2, 0}}.range();
    auto b = TAT::Tensor<double, TAT::NoSymmetry>{{"C", "D"}, {0, 2}}.range();
    auto c = TAT::Tensor<double, TAT::NoSymmetry>::contract(a, b, {{"B", "C"}});
    ASSERT_NE(c.storage().size(), 0);
    ASSERT_FLOAT_EQ(c.norm<-1>(), 0);
}

TEST(test_contract, corner_z2_symmetry_0k) {
    auto a = TAT::Tensor<double, TAT::Z2Symmetry>{{"A", "B"}, {{{0, 2}}, {{0, 0}}}}.range();
    auto b = TAT::Tensor<double, TAT::Z2Symmetry>{{"C", "D"}, {{{0, 0}}, {{0, 2}}}}.range();
    auto c = TAT::Tensor<double, TAT::Z2Symmetry>::contract(a, b, {{"B", "C"}});
    ASSERT_NE(c.storage().size(), 0);
    ASSERT_FLOAT_EQ(c.norm<-1>(), 0);
}

TEST(test_contract, corner_z2_symmetry_not_match_missing_left) {
    auto a = TAT::Tensor<double, TAT::Z2Symmetry>{{"A", "B"}, {{{1, 2}}, {{0, 2}}}}.range();
    auto b = TAT::Tensor<double, TAT::Z2Symmetry>{{"C", "D"}, {{{0, 2}}, {{0, 2}}}}.range();
    auto c = TAT::Tensor<double, TAT::Z2Symmetry>::contract(a, b, {{"B", "C"}});
    ASSERT_FLOAT_EQ(c.norm<0>(), 0);
}

TEST(test_contract, corner_z2_symmetry_not_match_missing_right) {
    auto a = TAT::Tensor<double, TAT::Z2Symmetry>{{"A", "B"}, {{{0, 2}}, {{0, 2}}}}.range();
    auto b = TAT::Tensor<double, TAT::Z2Symmetry>{{"C", "D"}, {{{0, 2}}, {{1, 2}}}}.range();
    auto c = TAT::Tensor<double, TAT::Z2Symmetry>::contract(a, b, {{"B", "C"}});
    ASSERT_FLOAT_EQ(c.norm<0>(), 0);
}

TEST(test_contract, corner_z2_symmetry_not_match_missing_middle) {
    auto a = TAT::Tensor<double, TAT::Z2Symmetry>{{"A", "B"}, {{{0, 2}}, {{1, 2}}}}.range();
    auto b = TAT::Tensor<double, TAT::Z2Symmetry>{{"C", "D"}, {{{1, 2}}, {{0, 2}}}}.range();
    auto c = TAT::Tensor<double, TAT::Z2Symmetry>::contract(a, b, {{"B", "C"}});
    ASSERT_NE(c.storage().size(), 0);
    ASSERT_FLOAT_EQ(c.norm<-1>(), 0);
}
