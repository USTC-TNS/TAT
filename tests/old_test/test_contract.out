{[32mnames[0m:[A,B],[32medges[0m:[2,2],[32mblocks[0m:[0,1,2,3]}
{[32mnames[0m:[C,D],[32medges[0m:[2,2],[32mblocks[0m:[0,1,2,3]}
{[32mnames[0m:[B,D],[32medges[0m:[2,2],[32mblocks[0m:[4,6,6,10]}
{[32mnames[0m:[B,C],[32medges[0m:[2,2],[32mblocks[0m:[2,6,3,11]}
{[32mnames[0m:[A,D],[32medges[0m:[2,2],[32mblocks[0m:[2,3,6,11]}
{[32mnames[0m:[A,C],[32medges[0m:[2,2],[32mblocks[0m:[1,3,3,13]}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[1,3,3,1],[32mblocks[0m:[316,796,1276,428,1164,1900,540,1532,2524]}
{[32mnames[0m:[A,B,C,D],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:1,1:2}},{arrow:0,segment:{0:2,1:2}},{arrow:1,segment:{0:2,-1:1,-2:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,1,2,3],[33m[-1,1,0,0][0m:[4,5,6,7,8,9,10,11],[33m[-1,1,1,-1][0m:[12,13,14,15],[33m[0,0,0,0][0m:[16,17,18,19],[33m[0,0,1,-1][0m:[20,21],[33m[0,1,0,-1][0m:[22,23,24,25],[33m[0,1,1,-2][0m:[26,27,28,29,30,31,32,33],[33m[-2,1,1,0][0m:[34,35,36,37,38,39,40,41]}}
{[32mnames[0m:[E,F,G,H],[32medges[0m:[{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}},{arrow:1,segment:{0:1,-1:2}},{arrow:0,segment:{0:2,1:1,2:2}}],[32mblocks[0m:{[33m[0,-2,0,2][0m:[0,1,2,3],[33m[0,-1,0,1][0m:[4,5],[33m[0,-1,-1,2][0m:[6,7,8,9,10,11,12,13],[33m[0,0,0,0][0m:[14,15,16,17,18,19,20,21],[33m[0,0,-1,1][0m:[22,23,24,25,26,27,28,29],[33m[1,-2,0,1][0m:[30],[33m[1,-2,-1,2][0m:[31,32,33,34],[33m[1,-1,0,0][0m:[35,36],[33m[1,-1,-1,1][0m:[37,38],[33m[1,0,-1,0][0m:[39,40,41,42,43,44,45,46]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
